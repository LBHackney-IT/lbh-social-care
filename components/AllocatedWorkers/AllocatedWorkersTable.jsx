import PropTypes from 'prop-types';

const AllocatedWorkersEntry = ({
  allocatedWorkerTeam,
  allocatedWorker,
  allocationStartDate,
  workerType,
  index,
}) => (
  <>
    <h3 className="govuk-fieldset__legend--m govuk-custom-text-color govuk-!-margin-top-0">
      Allocated Worker {index + 1}
    </h3>
    <hr className="govuk-divider" />
    <dl className="govuk-summary-list">
      {allocatedWorker && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Allocated Worker:</dt>
          <dd className="govuk-summary-list__value">{allocatedWorker}</dd>
        </div>
      )}
      {allocatedWorkerTeam && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Allocated Team</dt>
          <dd className="govuk-summary-list__value">{allocatedWorkerTeam}</dd>
        </div>
      )}
      {workerType && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Role</dt>
          <dd className="govuk-summary-list__value">{workerType}</dd>
        </div>
      )}
      {allocationStartDate && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Start Date</dt>
          <dd className="govuk-summary-list__value">
            {new Date(allocationStartDate).toLocaleDateString('en-GB')}
          </dd>
        </div>
      )}
    </dl>
  </>
);

const AllocatedWorkersTable = ({ records }) => (
  <table className="govuk-table">
    <tbody className="govuk-table__body">
      {records.map((result, index) => (
        <AllocatedWorkersEntry
          key={result.personId}
          index={index}
          {...result}
        />
      ))}
    </tbody>
  </table>
);

AllocatedWorkersTable.propTypes = {
  records: PropTypes.arrayOf(
    PropTypes.shape({
      allocatedWorker: PropTypes.string,
      role: PropTypes.string,
      allocatedTeam: PropTypes.string,
      startDate: PropTypes.number,
    })
  ).isRequired,
};

export default AllocatedWorkersTable;
