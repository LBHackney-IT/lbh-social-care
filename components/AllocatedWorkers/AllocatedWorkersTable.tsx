import { useRouter } from 'next/router';
import Link from 'next/link';
import { Allocation } from 'types';

interface EntryProps extends Allocation {
  deallocationUrl: string;
  index: number;
  showDeallocateButton: boolean;
}

const AllocatedWorkersEntry = ({
  allocatedWorkerTeam,
  allocatedWorker,
  allocationStartDate,
  allocationEndDate,
  workerType,
  showDeallocateButton,
  deallocationUrl,
}: EntryProps): React.ReactElement => (
  <section>
    <div>
      <h3>{allocatedWorker}</h3>
      {showDeallocateButton && (
        <Link href={deallocationUrl}>
          <a className="lbh-link">Deallocate</a>
        </Link>
      )}
    </div>

    <dl className="govuk-summary-list">
      {allocatedWorkerTeam && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Team</dt>
          <dd className="govuk-summary-list__value">{allocatedWorkerTeam}</dd>
        </div>
      )}
      {workerType && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Role</dt>
          <dd className="govuk-summary-list__value">{workerType}</dd>
        </div>
      )}
      {allocationStartDate && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">Start date</dt>
          <dd className="govuk-summary-list__value">
            {new Date(allocationStartDate).toLocaleDateString('en-GB')}
          </dd>
        </div>
      )}
      {allocationEndDate && (
        <div className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">End date</dt>
          <dd className="govuk-summary-list__value">
            {new Date(allocationEndDate).toLocaleDateString('en-GB')}
          </dd>
        </div>
      )}
    </dl>
  </section>
);

export interface Props {
  records: Allocation[];
  hasAllocationsPermissions: boolean;
}

const AllocatedWorkersTable = ({
  records,
  hasAllocationsPermissions,
}: Props): React.ReactElement => {
  const { asPath } = useRouter();

  return (
    <div>
      {records.map((record, index) => (
        <AllocatedWorkersEntry
          key={record.id}
          index={index}
          showDeallocateButton={hasAllocationsPermissions}
          deallocationUrl={`${asPath}/allocations/${record.id}/remove`}
          {...record}
        />
      ))}
    </div>
  );
};

export default AllocatedWorkersTable;
