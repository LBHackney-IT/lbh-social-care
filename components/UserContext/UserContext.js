import { createContext, useContext, useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { AUTH_WHITELIST } from 'utils/auth';
import { getUser } from 'utils/api/checkAuth';
import Layout from 'components/Layout';
import Spinner from 'components/Spinner/Spinner';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState();
  const [redirect, setRedirect] = useState();
  const [isLoading, setLoading] = useState(true);
  const router = useRouter();
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        const data = getUser();
        setUser(data);
      } catch (e) {
        setUser();
        setRedirect(e.response.status === 403 ? '/access-denied' : '/login');
      }
      setLoading(false);
    };
    initializeAuth();
  }, []);
  if (isLoading || !router.isReady) {
    return (
      <Layout>
        <Spinner />
      </Layout>
    );
  }
  if (
    typeof window !== 'undefined' &&
    AUTH_WHITELIST.includes(router.pathname) === Boolean(user)
  ) {
    router.push(redirect || location);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
        isLoading,
        setUser,
        redirect,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
