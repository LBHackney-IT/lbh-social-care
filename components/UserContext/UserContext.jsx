import { createContext, useContext } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

export const UserContext = createContext({
  user: undefined,
  setUser: () => {},
});

export const AuthProvider = ({ children, user }) => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export function useAuth() {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}
