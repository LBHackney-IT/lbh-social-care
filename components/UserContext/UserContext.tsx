import { createContext, useContext, ReactNode, ReactElement } from 'react';
import { useRouter } from 'next/router';

import { shouldRedirect } from 'utils/auth';
import { isBrowser } from 'utils/ssr';

import type { User } from 'types';

interface Props {
  children: ReactNode;
  user?: User;
}

export const UserContext = createContext<{ user?: User }>({
  user: undefined,
});

export const AuthProvider = ({ children, user }: Props): ReactElement => {
  const { pathname, push } = useRouter();
  const redirect = isBrowser() && shouldRedirect(pathname, user);
  if (redirect) {
    push(redirect);
    return <></>;
  }
  return (
    <UserContext.Provider
      value={{
        user,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};

export const useAuth = (): { user?: User } => {
  const context = useContext(UserContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};
