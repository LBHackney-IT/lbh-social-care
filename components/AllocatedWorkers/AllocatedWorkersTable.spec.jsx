import { render } from '@testing-library/react';
import { UserContext } from 'components/UserContext/UserContext';
import AllocatedWorkersTable from './AllocatedWorkersTable';

describe('AllocatedWorkers component', () => {
  const props = {
    records: [
      {
        allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
        allocatedWorker: 'Officer Name',
        allocationEndDate: '2019-03-28 00:00:00',
        allocationStartDate: '2019-03-28 00:00:00',
        workerType: 'Consultant Social Worker',
        id: 123,
        personId: 1234,
      },
    ],
    updateWorkers: jest.fn(),
  };
  it('should render properly', () => {
    const { asFragment } = render(
      <UserContext.Provider
        value={{
          user: { name: 'foo' },
        }}
      >
        <AllocatedWorkersTable {...props} />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
