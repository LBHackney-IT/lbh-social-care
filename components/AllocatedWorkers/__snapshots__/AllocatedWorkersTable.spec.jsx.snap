// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllocatedWorkers component should render properly - with deallocate button 1`] = `
<DocumentFragment>
  <div>
    <div
      class="lbh-table-header"
    >
      <h3
        class="govuk-fieldset__legend--m govuk-custom-text-color govuk-!-margin-top-0"
      >
        ALLOCATED WORKER 1
      </h3>
      <button
        class="govuk-button govuk-button--secondary"
        data-module="govuk-button"
      >
        Deallocate Worker
      </button>
    </div>
    <hr
      class="govuk-divider"
    />
    <div />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Allocated worker:
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Officer Name
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Allocated team
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Safeguarding and Reviewing Team
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Role
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Consultant Social Worker
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Start date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          28/03/2019
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          End date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          28/03/2019
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;

exports[`AllocatedWorkers component should render properly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="lbh-table-header"
    >
      <h3
        class="govuk-fieldset__legend--m govuk-custom-text-color govuk-!-margin-top-0"
      >
        ALLOCATED WORKER 1
      </h3>
    </div>
    <hr
      class="govuk-divider"
    />
    <div />
    <dl
      class="govuk-summary-list"
    >
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Allocated worker:
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Officer Name
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Allocated team
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Safeguarding and Reviewing Team
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Role
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          Consultant Social Worker
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          Start date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          28/03/2019
        </dd>
      </div>
      <div
        class="govuk-summary-list__row"
      >
        <dt
          class="govuk-summary-list__key"
        >
          End date
        </dt>
        <dd
          class="govuk-summary-list__value"
        >
          28/03/2019
        </dd>
      </div>
    </dl>
  </div>
</DocumentFragment>
`;
